pkg/controllers/namespace/namespace_sync_controller_test.go

```diff
@@ -0,0 +1,505 @@
+			expectedError:  false,
+		},
+		{
+			name: "Namespace with skip auto propagation label",
+			namespace: &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "skip-namespace",
+					Labels: map[string]string{
+						policyv1alpha1.NamespaceSkipAutoPropagationLabel: "true",
+					},
+				},
+			},
+			expectedResult: controllerruntime.Result{},
+			expectedError:  false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(tt.namespace).WithLists(&clusterv1alpha1.ClusterList{Items: tt.clusters}).Build()
+			fakeRecorder := record.NewFakeRecorder(100)
+
+			c := &Controller{
+				Client:          fakeClient,
+				EventRecorder:   fakeRecorder,
+				OverrideManager: overridemanager.New(fakeClient, fakeRecorder),
+			}
+
+			req := controllerruntime.Request{
+				NamespacedName: types.NamespacedName{
+					Name: tt.namespace.Name,
+				},
+			}
+
+			result, err := c.Reconcile(context.Background(), req)
+			assert.Equal(t, tt.expectedResult, result)
+			assert.Equal(t, tt.expectedError, err != nil)
+
+			if tt.name == "Namespace should be synced" {
+				work := &workv1alpha1.Work{}
+				err = fakeClient.Get(context.Background(), types.NamespacedName{
+					Namespace: names.GenerateExecutionSpaceName(tt.clusters[0].Name),
+					Name:      names.GenerateWorkName(tt.namespace.Kind, tt.namespace.Name, tt.namespace.Namespace),
+				}, work)
+				assert.NoError(t, err)
+				assert.NotNil(t, work)
+			}
+		})
+	}
+}
+
+func TestController_buildWorks(t *testing.T) {
+	scheme := runtime.NewScheme()
+	_ = corev1.AddToScheme(scheme)
+	_ = clusterv1alpha1.Install(scheme)
+	_ = workv1alpha1.Install(scheme)
+	_ = policyv1alpha1.Install(scheme)
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-namespace",
+		},
+	}
+
+	clusters := []clusterv1alpha1.Cluster{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cluster1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cluster2",
+			},
+		},
+	}
+
+	fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(namespace, &clusters[0], &clusters[1]).Build()
+	fakeRecorder := record.NewFakeRecorder(100)
+
+	c := &Controller{
+		Client:          fakeClient,
+		EventRecorder:   fakeRecorder,
+		OverrideManager: overridemanager.New(fakeClient, fakeRecorder),
+	}
+
+	err := c.buildWorks(context.Background(), namespace, clusters)
+	assert.NoError(t, err)
+
+	for _, cluster := range clusters {
+		work := &workv1alpha1.Work{}
+		err = fakeClient.Get(context.Background(), types.NamespacedName{
+			Namespace: names.GenerateExecutionSpaceName(cluster.Name),
+			Name:      names.GenerateWorkName(namespace.Kind, namespace.Name, namespace.Namespace),
+		}, work)
+		assert.NoError(t, err)
+		assert.NotNil(t, work)
+		assert.Equal(t, namespace.Name, work.OwnerReferences[0].Name)
+	}
+}
+
+func TestController_buildWorksWithOverridePolicy(t *testing.T) {
```

Comment:
Is it convenient to combine this test method with TestController_buildWorks into one test method?

Reply:
I think separate tests will ensure that override policy behavior doesn't interfere with basic functionality. I could create helper functions for common setup code if you say.

Reply:
You can modify it to see the effect.